File: openadmin/spec/table.py
# SPDX-FileCopyrightText: 2026 OpenAdmin
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from collections.abc import Iterable
from typing import Any, Literal, NotRequired

from typing_extensions import TypedDict

from .colors import Color
from .http_methods import HttpMethod
from .icons import Icon
from .json_schema import JsonSchema

type ColumnStyle = Literal["avatar", "image", "badge", "link"]


class ColumnConfigValue(TypedDict):
    style: NotRequired[ColumnStyle]
    label: NotRequired[str]
    icon: NotRequired[Icon]
    color: NotRequired[Color]


class ActionConfig(TypedDict):
    action: str
    label: NotRequired[str]
    icon: NotRequired[Icon]
    color: NotRequired[Color]

    query: dict[str, Any]
    body: dict[str, Any]
    form: dict[str, Any]


class TableComponent(TypedDict):
    type: Literal["table"]
    id: str
    name: str
    description: str | None
    columns: dict[str, ColumnConfigValue] | None
    icon: Icon | None
    color: Color | None
    method: HttpMethod
    is_hidden: bool
    refresh: int | None
    form: JsonSchema | None
    body: JsonSchema | None
    query: JsonSchema | None


TableData = Iterable[
    TypedDict(
        "TableRow",
        {
            "__view__": str | int | float | bool | None,
            "__actions__": list[ActionConfig],
            "__style__": ColumnStyle | None,
        },
        extra_items=str | int | float | bool | None,
    )
    | object
]


class TableResponse(TypedDict):
    data: TableData
    icon: NotRequired[Icon]
    color: NotRequired[Color]
    refresh: NotRequired[int | None]


type Table = TableData | TableResponse

File: client/src/schemas/table.ts
// SPDX-FileCopyrightText: 2026 OpenAdmin
//
// SPDX-License-Identifier: AGPL-3.0-or-later

import { z } from "zod"
import { colorSchema } from "./color"
import { httpMethodSchema } from "./http-method"
import { iconSchema } from "./icon"
import { jsonSchemaSchema } from "./json-schema"

export const COLUMN_STYLES = ["avatar", "image", "badge", "link"] as const

export const columnStyleSchema = z.enum(COLUMN_STYLES)

export type ColumnStyle = z.infer<typeof columnStyleSchema>

export const columnConfigValueSchema = z.object({
	style: columnStyleSchema.optional(),
	label: z.string().optional(),
	icon: iconSchema.optional(),
	color: colorSchema.optional(),
})

export type ColumnConfigValue = z.infer<typeof columnConfigValueSchema>

export const actionConfigSchema = z.object({
	action: z.string(),
	label: z.string().optional(),
	icon: iconSchema.optional(),
	color: colorSchema.optional(),
	query: z.record(z.string(), z.unknown()),
	body: z.record(z.string(), z.unknown()),
	form: z.record(z.string(), z.unknown()),
})

export type ActionConfig = z.infer<typeof actionConfigSchema>

export const tableComponentSchema = z.object({
	type: z.literal("table"),
	id: z.string(),
	name: z.string(),
	description: z.string().nullable(),
	columns: z.record(z.string(), columnConfigValueSchema).nullable(),
	icon: iconSchema.nullable(),
	color: colorSchema.nullable(),
	method: httpMethodSchema,
	is_hidden: z.boolean(),
	refresh: z.number().nullable(),
	form: jsonSchemaSchema.nullable(),
	body: jsonSchemaSchema.nullable(),
	query: jsonSchemaSchema.nullable(),
})

export type TableComponent = z.infer<typeof tableComponentSchema>

const tableRowValueSchema = z.union([z.string(), z.number(), z.boolean(), z.null()])

const tableRowSchema = z
	.object({
		__view__: tableRowValueSchema.optional(),
		__actions__: z.array(actionConfigSchema).optional(),
		__style__: columnStyleSchema.nullable().optional(),
	})
	.catchall(tableRowValueSchema)

export const tableDataSchema = z.array(z.union([tableRowSchema, z.unknown()]))

export type TableData = z.infer<typeof tableDataSchema>

export const tableResponseSchema = z.object({
	data: tableDataSchema,
	icon: iconSchema.optional(),
	color: colorSchema.optional(),
	refresh: z.number().nullable().optional(),
	total: z.number().optional(),
})

export type TableResponse = z.infer<typeof tableResponseSchema>

export const tableSchema = z.union([tableDataSchema, tableResponseSchema])

export type Table = z.infer<typeof tableSchema>
